include/unifex/when_all_range.hpp

```diff
@@ -0,0 +1,341 @@
+ *
+ *   https://llvm.org/LICENSE.txt
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#pragma once
+
+#include <algorithm>
+#include <atomic>
+#include <cstddef>
+#include <functional>
+#include <type_traits>
+#include <vector>
+
+#include <unifex/blocking.hpp>
+#include <unifex/get_stop_token.hpp>
+#include <unifex/inplace_stop_token.hpp>
+#include <unifex/manual_lifetime.hpp>
+#include <unifex/optional.hpp>
+#include <unifex/receiver_concepts.hpp>
+#include <unifex/sender_concepts.hpp>
+#include <unifex/std_concepts.hpp>
+#include <unifex/type_list.hpp>
+#include <unifex/utility.hpp>
+#include <unifex/variant.hpp>
+
+#include <unifex/detail/prologue.hpp>
+
+namespace unifex {
+namespace _when_all_range {
+
+template <typename Receiver, typename Sender>
+struct _operation {
+  struct type;
+};
+
+template <typename Receiver, typename Sender>
+struct _element_receiver {
+  struct type;
+};
+
+template <typename Receiver, typename Sender>
+using element_receiver = typename _element_receiver<Receiver, Sender>::type;
+
+template <typename Receiver, typename Sender>
+using operation =
+    typename _operation<unifex::remove_cvref_t<Receiver>, Sender>::type;
+
+template <typename Receiver, typename Sender>
+struct _operation<Receiver, Sender>::type final {
+private:
+  using sender_nonvoid_value_type =
+      unifex::sender_single_value_result_t<unifex::remove_cvref_t<Sender>>;
+
+  using element_receiver_t = element_receiver<Receiver, Sender>;
+
+  friend struct _element_receiver<Receiver, Sender>::type;
+
+  struct cancel_operation final {
+    type& op_;
+    void operator()() noexcept { op_.request_stop(); }
+  };
+
+  struct _operation_holder final {
+    unifex::optional<sender_nonvoid_value_type> value;
+    unifex::connect_result_t<Sender, element_receiver_t> connection;
+
+    _operation_holder(Sender&& sender, type& op, std::size_t index) noexcept(
+        unifex::is_nothrow_connectable_v<Sender, element_receiver_t>&& std::
+            is_nothrow_constructible_v<element_receiver_t, type&, std::size_t>)
+      : connection(unifex::connect(
+            static_cast<Sender&&>(sender), element_receiver_t(op, index))) {}
+  };
+
+public:
+  type(Receiver&& receiver, std::vector<Sender>&& senders)
+    : receiver_(std::move(receiver))
+    , refCount_(senders.size()) {
+    std::allocator<_operation_holder> allocator;
+    holders_ = allocator.allocate(senders.size());
+    try {
+      for (auto&& sender : senders) {
+        new (holders_ + numHolders_)
+            _operation_holder{std::move(sender), *this, numHolders_};
+        ++numHolders_;
+      }
+    } catch (...) {
+      std::destroy(holders_, holders_ + numHolders_);
```

Comment:
I wonder if it's worth using `std::make_reverse_iterator` to destroy things in reverse order…
```suggestion
      std::destroy(
          std::make_reverse_iterator(holders_ + numHolders_),
          std::make_reverse_iterator(holders_));
```